common/lib/dependabot/file_parsers/base/dependency_set.rb

```diff
@@ -14,34 +14,44 @@ def initialize(dependencies = [], case_sensitive: false)
             raise ArgumentError, "must be an array of Dependency objects"
           end
 
-          @dependencies = dependencies
           @case_sensitive = case_sensitive
+          @dependencies = Hash.new { |hsh, key| hsh[key] = DependencySlot.new }
+          dependencies.each { |dep| self << dep }
         end
 
-        attr_reader :dependencies
+        def dependencies
+          @dependencies.values.filter_map(&:combined)
+        end
 
         def <<(dep)
           raise ArgumentError, "must be a Dependency object" unless dep.is_a?(Dependency)
 
-          existing_dependency = dependency_for_name(dep.name)
+          @dependencies[key_for_dependency(dep)] << dep
+          self
+        end
 
-          return self if existing_dependency&.to_h == dep.to_h
+        def +(other)
+          raise ArgumentError, "must be a DependencySet" unless other.is_a?(DependencySet)
 
-          if existing_dependency
-            dependencies[dependencies.index(existing_dependency)] =
-              combined_dependency(existing_dependency, dep)
-          else
-            dependencies << dep
+          other_names = other.dependencies.map(&:name)
+          other_names.each do |name|
+            all_versions = other.all_versions_for_name(name)
+            all_versions.each { |dep| self << dep }
           end
 
           self
         end
 
-        def +(other)
-          raise ArgumentError, "must be a DependencySet" unless other.is_a?(DependencySet)
+        def all_versions_for_name(name, sort: false)
```

Comment:
Now that I'm looking at it, this still calls `#sort!` rather than `#sort` because the first pass wanted to mutate the array. This is a bug but it hasn't come up yet since nothing is using this code and is more than enough to convince me to delete the code!